### PR TITLE
Add golden tests for generated files

### LINKED ISSUE #50: Add golden tests for cookiecutter
A very easy way to make sure templates don't introduce regressions is to do [golden testing](https://en.wikipedia.org/wiki/Characterization_test).

This basically consists in testing the generated files manually and once they are verified to work well, store them as the "golden copy". Golden test consist in just comparing newly generated files with the golden copies.

This is an easy way to test complex outputs (like in this case, where testing involves running github actions, checking if owners are assigned to PRs, etc.) and making sure regressions are not introduced in refactoring for example.

There are 2 main golden tests libraries in Python:

* ~https://pypi.org/project/pytest-goldie/ | [snyk report](https://snyk.io/advisor/python/pytest-goldie)~
* ~https://pypi.org/project/pytest-golden/ | [snyk report](https://snyk.io/advisor/python/pytest-golden)~

Both compare similarly in terms of usage and sustainability according to [Snyk](https://snyk.io/advisor), so we should compare the features and general feel to select one.

**Update:** It looks like  `pytest-goldie` might be a fake. The source is not available in pypi, the home page gives a 404.

### PR DESCRIPTION
**Pull request by llucax**
Testing the generated templates are correct is hard. For now we are testing that the basics work by running `nox` on some generated projects but this will not test bugs in documentation, comments, etc. Nor will test GitHub Actions workflows or other advanced configuration.

By adding golden tests we at least ensure that once the results are manually tested, we won't introduce unexpected regressions.

We considered using the pytest-golden plugin, but unfortunately it does not support using whole files and directory trees as goldens, so we built our own ad-hoc solution for now.

To update the golden files the test file needs to be temporarily updated to set `UPDATE_GOLDEN` to `True` to replace the golden files with the newly generated files by running `pytest` again, which is not great, but it is good enough for now.

Fixes #50.

### CODE REVIEW

**Comment by llucax** [comment]
Setting auto-merge.

**Comment by llucax** [comment]
Damn, it looks like this is failing because `git` output changes from version to version. I'll need to figure out something to deal with this.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -60,6 +60,36 @@ nox -R -s pylint -- test/test_*.py
...
+   so that the generated files match the golden files again.
+
+2. The generated files don't match the golden files because an intended change
+   was introduced. In this case, the golden files need to be updated.
+
+In the latter case, manually updating files is complicated and error-prone, so
+a simpler (though hacky) way is provided.
+
+To update the golden files, simply edit the
+`tests/integration/test_cookiecutter_generation.py` file and temporarily set
```
[the remark this answers is not in the document]
Comment (Marenz) [reply CONTRIBUTING.md:85]: I am just thinking, changing a flag involves opening the file, navigating the curser etc, where as a simple script can just be run and done :D 

Either way fine by me

[the remark this answers is not in the document]
Comment (llucax) [reply CONTRIBUTING.md:85]: Yes, I understand that running a script is simpler, but to separate the logic to regenerate the files things get more complicated, if we just repeat the commands used to generate the files in a separate script there is the risk that they get out of sync. All golden testing libraries I used work like this, you just run the tests with an extra flag like `--update-golden`, I don't think having a separate script is moving in the right direction.

We could investigate if we can add pytest CLI flags inside a test or something, but again IMHO is too much time investment for this.

**Comment by llucax** [comment]
OK, this was updated with quite a few changes to get rid of the flakiness that made the tests pass locally but not in the CI. Can you give it another look?

**Comment by llucax** [comment]
One more fix coming up soon (hopefully the last one :unamused: )

**Comment by llucax** [comment]
Updated, should be ready for the final review.

**Review by Marenz — DISMISSED** [review]

**Comment by llucax** [comment]
Rebased, so it needs a new approve @Marenz

**Review by Marenz — APPROVED** [review]

**Pull request merged**